src/components/ModelingSidebar/ModelingPanes/CodeEditor.tsx

```diff
@@ -0,0 +1,140 @@
+import React, { useEffect, useMemo, useRef, useState } from 'react'
+import {
+  EditorState,
+  EditorStateConfig,
+  Extension,
+  StateEffect,
+} from '@codemirror/state'
+import { EditorView } from '@codemirror/view'
+import { oneDark } from '@codemirror/theme-one-dark'
+
+//reference: https://github.com/sachinraja/rodemirror/blob/main/src/use-first-render.ts
+const useFirstRender = () => {
+  const firstRender = useRef(true)
+
+  useEffect(() => {
+    firstRender.current = false
+  }, [])
+
+  return firstRender.current
+}
+
+const defaultLightThemeOption = EditorView.theme(
+  {
+    '&': {
+      backgroundColor: '#fff',
+    },
+  },
+  {
+    dark: false,
+  }
+)
+
+interface ICodeEditor {
+  onView?: (view: EditorView | null) => void
+  initialDocValue?: EditorStateConfig['doc']
+  extensions?: Extension
+  theme: 'light' | 'dark'
+  autoFocus?: boolean
+  selection?: EditorStateConfig['selection']
+}
+
+interface UseCodeMirror extends ICodeEditor {
+  container?: HTMLDivElement | null
+}
+
+const CodeEditor: React.FC<ICodeEditor> = ({
+  onView,
+  extensions = [],
+  initialDocValue,
+  theme,
+  autoFocus = false,
+  selection,
+}) => {
+  const editorRef = useRef<HTMLDivElement>(null)
+
+  useCodeMirror({
+    container: editorRef.current,
+    onView,
+    extensions,
+    initialDocValue,
+    theme,
+    autoFocus,
+    selection,
+  })
+
+  return <section ref={editorRef}></section>
+}
+
+export function useCodeMirror(props: UseCodeMirror) {
+  const {
+    onView,
+    extensions = [],
+    initialDocValue,
+    theme,
+    autoFocus = false,
+    selection,
+  } = props
+
+  const [container, setContainer] = useState<HTMLDivElement | null>()
+  const [editorView, setEditorView] = useState<EditorView | null>(null)
+
+  const isFirstRender = useFirstRender()
+
+  const targetExtensions = useMemo(() => {
```

Comment:
nit: I would avoid stuff like useMemo or useCallback until you're certain these are causing perf issues - too many times I've seen people hit update issues around them

Reply:
I def know we don't want this to change unless the extensions changed if thats what you mean

Reply:
useEffect should be used instead with a useState but honestly this is fine too. It pretty much equates to the same thing.

Reply:
useMemo here just prevents a re-computation so it's an optimization the original did